File: DigitalImageProcessing/cpp/erode_dilate_image.cpp
/**
 * @File    : erode_dilate_image.cpp
 * @Brief   : 形态学操作：腐蚀(选择一个任意领域[结构元]里面的最小值)和膨胀(选择一个任意领域[结构元]里面的最大值)
 * @Author  : Wei Li
 * @Date    : 2021-10-11
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

// 2. 腐蚀操作
int radio = 1;     // 结构元半径
int Max_R = 20;   // 设置最大半径
cv::Mat d;       // 膨胀后图像
cv::Mat image;  // 输入图像

// 回调函数，调节 radio
void callBackRadio(int, void*)
{
    // 创建只有垂直方向的矩形结构元
    cv::Mat s = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(1, 2*radio+1));
    // 进行膨胀操作
    cv::dilate(image, d, s);
    cv::imshow("DilateImage", d);
}

// -------------------------------
int main(int argc, char **argv)
{
    image = cv::imread(argv[1], 0);
    if (!image.data)
    {
        std::cout << "Error: Reading image file unsuccessfully.";
        return -1;
    }
    cv::imshow("OriginImage", image);

    // 1. 腐蚀操作
    // 创建结构元
    cv::Mat s = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
    cv::Mat e;
    // 腐蚀操作，迭代次数(腐蚀次数) 2
    cv::erode(image, e, s, cv::Point(-1, -1), 2);
    cv::imshow("ErodeImage", e);

    // 2. 腐蚀操作
    // 创建显示膨胀效果的窗口
    cv::namedWindow("DilateWin", 1);
    // 创建调节 半径 radio 的进度条
    cv::createTrackbar("Radio", "DilateWin", &radio, Max_R, callBackRadio);
    callBackRadio(0, 0);

    cv::waitKey(0);
    cv::destroyAllWindows();
    return 0;
}

File: DigitalImageProcessing/python/erode_dilate_image.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 形态学操作：腐蚀(选择一个任意领域[结构元]里面的最小值)和膨胀(选择一个任意领域[结构元]里面的最大值)
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-11
"""

import sys

import cv2 as cv


# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or dictory.")
        cv.imshow("OriginImage",image)

        # 1. 腐蚀操作
        # 创建结构元——矩形
        s = cv.getStructuringElement(cv.MORPH_RECT, (3, 3))
        # 腐蚀图像，迭代次数为 1
        r = cv.erode(image, s)
        cv.imshow("ErodeImage", r)
        # 边界提取
        e = image - r
        cv.imshow("EdgeImage", e)

        # 2. 膨胀操作
        # 结构元半径
        radio = 1
        MAX_R = 20
        # 显示膨胀效果的窗口
        cv.namedWindow("dilateWin", 1)
        def nothing(*arg):
            pass
        # 调节结构云半径
        cv.createTrackbar("r", "dilateWin", radio, MAX_R, nothing)
        while True:
            # 获取当前的 radio 值
            radio = cv.getTrackbarPos("r", "dilateWin")
            # 创建结构元
            s_dilate = cv.getStructuringElement(cv.MORPH_ELLIPSE, (2*radio+1, 2*radio+1))
            # 膨胀后图像
            d = cv.dilate(image, s_dilate)
            cv.imshow("DilateImage", d)
            ch = cv.waitKey(5)
            # Esc 退出循环
            if ch == 27:
                break

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print("Usge: python.py imageFile")
